### Домашнее задание к лекции  "Проблема качества данных"
Необходимо запустить практическую часть занятия, и посмотреть самому то, о чём говорили на лекции.   По образу практики, попробуйте создать искусственный датасет с лишними столбцами.   Целевую метку, при правильной обработке данных, формируйте таким образом, чтобы без затруднений её смогла описать линейная модель.   Ориентируйтесь на то, что было показано во время занятия, и каждый шаг описывайте в markdown.   Здесь важно видеть ваш ход мысли.   Не бойтесь ошибиться или написать не то. Данное задание не имеет какого-то “правильного” решения. Цель - достичь базового понимания проблемы. Чем больше вы фантазируете, тем лучше :) Тем не менее, старайтесь представить те ситуации, которые по-вашему мнению могли бы быть в реальных данных. Успеха!

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression

In [4]:
# Создаю синтетически данные по стоимости ковра в зависимости от размеров и времени его использования
n_samples = 1000

age_using = np.random.choice(90, n_samples) + 21
length = np.random.choice(100, n_samples) + 42
width = np.random.choice(110, n_samples) + 84

price = length * width * 100 + 42

data = pd.DataFrame({'age_using': age_using, 'length': length, 'width': width, 'price': price})
data.head(5)

age_using  length  width    price
0         70      65     90   585042
1         72      43    119   511742
2         89      79    153  1208742
3         22      95    153  1453542
4         42      67    153  1025142

In [5]:
# строю предсказание на основе линейной регрессии и вывожу ошибку 
from sklearn.metrics import mean_absolute_error

In [9]:
X = data[['age_using', 'length', 'width']]
y = data['price']
reg = LinearRegression().fit(X, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['age_using', 'length', 'width']])
print('Error : {}'.format(mean_absolute_error(pred_values, y)))
# получили очень большую ошибку

Weights: [ -141.69292931 13998.34095321  9019.44095983]
Bias: -1256997.1128383912
Error : 68925.10166674572


In [10]:
y.median()

1190242.0

In [11]:
# убираю признак по времени использования ковра
# в результате качество незначительно улучшилось
X = data[['length', 'width']]
y = data['price']
reg = LinearRegression().fit(X, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['length', 'width']])
print('Error: {}'.format(mean_absolute_error(pred_values, y)))

Weights: [13997.43812978  9028.40612888]
Bias: -1267462.436451451
Error: 68896.76243207353


In [12]:
# создаю новый признак площади
data['mult'] = data['length'] * data['width']
data.head(5)

age_using  length  width    price   mult
0         70      65     90   585042   5850
1         72      43    119   511742   5117
2         89      79    153  1208742  12087
3         22      95    153  1453542  14535
4         42      67    153  1025142  10251

In [13]:
# в результате практически нулевая ошибка со сдвигом 42
X = data[['mult']]
y = data['price']
reg = LinearRegression().fit(X, y)
print('Weights: {}'.format(reg.coef_))
print('Bias: {}'.format(reg.intercept_))

pred_values = reg.predict(data[['mult']])
print('Error: {}'.format(mean_absolute_error(pred_values, y)))

Weights: [100.]
Bias: 42.00000000023283
Error: 4.604225978255272e-11


Фактически мы пришли к изначальной формуле:  
  price = length * width * 100 + 42  
  где:  
  Weights: [100.] - вес(W)  
  Bias: 42.00000000023283 - сдвиг
  
  

In [ ]:
# в принципе ход мыслей понятен